pkg/cli/image/manifest/source.go

```diff
@@ -0,0 +1,134 @@
+package manifest
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/spf13/pflag"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	kcmdutil "k8s.io/kubectl/pkg/cmd/util"
+
+	operatorv1alpha1 "github.com/openshift/api/operator/v1alpha1"
+	operatorv1alpha1scheme "github.com/openshift/client-go/operator/clientset/versioned/scheme"
+	operatorv1alpha1client "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1alpha1"
+	"github.com/openshift/library-go/pkg/image/reference"
+)
+
+// ImageSourceOptions sets correct image source for oc adm commands
+type ImageSourceOptions struct {
+	ImageContentSourcePolicyFile string
+	ImageContentSourcePolicyList []operatorv1alpha1.ImageContentSourcePolicy
+}
+
+// Bind adds the options to the flag set.
+func (o *ImageSourceOptions) Bind(flags *pflag.FlagSet) {
+	flags.StringVar(&o.ImageContentSourcePolicyFile, "icsp-file", o.ImageContentSourcePolicyFile, "Path to an ImageContentSourcePolicy file.  If set, data from this file will be used to set source release image.")
+}
+
+// GetICSPs returns a list of ImageContentSourcePolicy objects found from file or from a connected cluster.
+func (o *ImageSourceOptions) GetICSPs(f kcmdutil.Factory) error {
+	// If ImageContentSourceFile is given, only add ImageContentSource from file, don't search cluster ICSP
+	if len(o.ImageContentSourcePolicyFile) != 0 {
+		icspData, err := ioutil.ReadFile(o.ImageContentSourcePolicyFile)
+		if err != nil {
+			return fmt.Errorf("unable to read ImageContentSourceFile %s: %v", o.ImageContentSourcePolicyFile, err)
+		}
+		if len(icspData) == 0 {
+			return fmt.Errorf("no data found in ImageContentSourceFile %s", o.ImageContentSourcePolicyFile)
+		}
+		icspObj, err := kruntime.Decode(operatorv1alpha1scheme.Codecs.UniversalDeserializer(), icspData)
+		if err != nil {
+			return fmt.Errorf("error decoding ImageContentSourcePolicy from %s: %v", o.ImageContentSourcePolicyFile, err)
+		}
+		var icsp *operatorv1alpha1.ImageContentSourcePolicy
+		var ok bool
+		if icsp, ok = icspObj.(*operatorv1alpha1.ImageContentSourcePolicy); !ok {
+			return fmt.Errorf("could not decode ImageContentSourcePolicy from %s", o.ImageContentSourcePolicyFile)
+		}
+		o.ImageContentSourcePolicyList = append(o.ImageContentSourcePolicyList, *icsp)
+		return nil
+	}
+	restConfig, err := f.ToRESTConfig()
```

Comment:
Unfortunately i think this structure probably won't work right - we don't want to make these calls here (which could be fairly expensive or slow), we want to only do them if we need them.  The check above is fine (since a user asked).

I think this method needs to provide a function that when asked can return alternates, that hides all of these details.  That function would be provided to the repository lookup client that could delegate to it late (and return an error or not).

I don't think doing this up front is going to scale, especially since this forces us to talk to a client in every image code path, even in places where we didn't do so before.

Reply:
got it, working on it, ok

Reply:
updated, now ImageSourceOptions has fn parameter, to pass an ICSPClient - and this client is _only_ used if the image lookup fails (via the function `GetICSPs(icspClient)`) (when using mirrored registries)